apiserver/pkg/manager/resource_manager.go

```diff
@@ -312,6 +313,35 @@ func (r *ResourceManager) CreateService(ctx context.Context, apiService *api.Ray
 	return newRayService, nil
 }
 
+func (r *ResourceManager) UpdateRayService(ctx context.Context, request *api.UpdateRayServiceRequest) (*v1alpha1.RayService, error) {
+	serviceName := request.Name
+	namespace := request.Namespace
+
+	client := r.getRayServiceClient(namespace)
+	service, err := getServiceByName(ctx, client, serviceName)
+	if err != nil {
+		return nil, util.Wrap(err, "Get service failure")
+	}
+	updateService := request.GetUpdateService()
+	// if workerGroupSpec is not nil, update worker group
+	if updateService.WorkerGroupUpdateSpec != nil {
+		oldWorkerGroups := service.Spec.RayClusterSpec.WorkerGroupSpecs
+		newWorkerGroups := util.UpdateRayServiceWorkerGroupSpecs(updateService.WorkerGroupUpdateSpec, oldWorkerGroups)
+		service.Spec.RayClusterSpec.WorkerGroupSpecs = newWorkerGroups
+	}
+	if updateService.ServeConfigs != nil {
+		oldServeConfigSpecs := service.Spec.ServeDeploymentGraphSpec.ServeConfigSpecs
+		newServeConfigSpecs := util.UpdateServeConfigSpecs(updateService.ServeConfigs, oldServeConfigSpecs)
+		service.Spec.ServeDeploymentGraphSpec.ServeConfigSpecs = newServeConfigSpecs
+	}
+
+	newService, err := client.Update(ctx, service, metav1.UpdateOptions{})
```

Comment:
so if `service` `resourceVersion` changes, the update could fail ?

Reply:
yeah, but k8s do update in a reconcile loop. So if it fails, it will retry in next loop. But this api call is 1 time call. Do we want to retry here? Or the caller should retry their side?